sql-plugin/src/main/scala/org/apache/spark/sql/rapids/stringFunctions.scala

```diff
@@ -1174,7 +1181,136 @@ case class GpuRegExpExtract(
+      case Literal(value, DataTypes.IntegerType) =>
+        val idx = value.asInstanceOf[Int]
+        if (idx < 0) {
+          willNotWorkOnGpu("the specified group index cannot be less than zero")
+        }
+        if (idx > numGroups) {
+          willNotWorkOnGpu(
+            s"regex group count is $numGroups, but the specified group index is $idx")
+        }
+      case _ =>
+        willNotWorkOnGpu("GPU only supports literal index")
+    }
+  }
+
+  override def convertToGpu(
+      str: Expression,
+      regexp: Expression,
+      idx: Expression): GpuExpression = {
+    val cudfPattern = pattern.getOrElse(
+      throw new IllegalStateException("Expression has not been tagged with cuDF regex pattern"))
+    GpuRegExpExtractAll(str, regexp, idx, numGroups, cudfPattern)
+  }
+}
+
+case class GpuRegExpExtractAll(
+    str: Expression,
+    regexp: Expression,
+    idx: Expression,
+    numGroups: Int,
+    cudfRegexPattern: String)
+  extends GpuRegExpTernaryBase with ImplicitCastInputTypes with NullIntolerant {
+
+  override def dataType: DataType = ArrayType(StringType, containsNull = true)
+  override def inputTypes: Seq[AbstractDataType] = Seq(StringType, StringType, IntegerType)
+  override def first: Expression = str
+  override def second: Expression = regexp
+  override def third: Expression = idx
+
+  override def prettyName: String = "regexp_extract_all"
 
+  override def doColumnar(
+      str: GpuColumnVector,
+      regexp: GpuScalar,
+      idx: GpuScalar): ColumnVector = {
+
+    idx.getValue.asInstanceOf[Int] match {
+      case 0 => 
+        str.getBase.extractAllRecord(cudfRegexPattern, 0)
+      case intIdx =>
+        // Extract matches corresponding to idx. cuDF's extract_all_record does not support 
+        // group idx, so we must manually extract the relevant matches. Example:
+        // Given the pattern (\d+)-(\d+) and idx=1
+        //
+        // |      Input      |      Java       |               cuDF             |
+        // |-----------------|-----------------|--------------------------------|
+        // | '1-2, 3-4, 5-6' | ['1', '3', '5'] | ['1', '2', '3', '4', '5', '6'] |
+        //
+        // Since idx=1 and the pattern has 2 capture groups, we take the 1st element and every
+        // 2nd element afterwards from the cuDF list
+
+        val rowCount = str.getRowCount
+        
+        val extractedWithNulls = withResource(
+          str.getBase.extractAllRecord(cudfRegexPattern, intIdx)) { allExtracted =>
+            withResource(allExtracted.countElements) { listSizes =>
+              withResource(listSizes.max) { maxSize =>
+                val maxSizeInt = maxSize.getInt
+                val stringCols: Seq[ColumnVector] = Range(intIdx - 1, maxSizeInt, numGroups).map { 
+                  i =>
+                    withResource(ColumnVector.fromScalar(Scalar.fromInt(i), rowCount.toInt)) { 
+                      index => allExtracted.extractListElement(index)
+                    }
```

Comment:
The scalar also needs to be wrapped with `withResource` here.

```suggestion
                    withResource(Scalar.fromInt(i)) { scalarIndex =>
                      withResource(ColumnVector.fromScalar(scalarIndex, rowCount.toInt)) {
                        index => allExtracted.extractListElement(index)
                      }
                    }
```